src/rtc.rs

```diff
@@ -0,0 +1,448 @@
+        let dr = self.regs.dr.read();
+        let day = bcd2_decode(dr.dt().bits(), dr.du().bits());
+        Ok(day as u8)
+    }
+
+    fn get_month(&mut self) -> Result<u8, Self::Error> {
+        let dr = self.regs.dr.read();
+        let mt: u8 = if dr.mt().bit() { 1 } else { 0 };
+        let month = bcd2_decode(mt, dr.mu().bits());
+        Ok(month as u8)
+    }
+
+    fn get_year(&mut self) -> Result<u16, Self::Error> {
+        let dr = self.regs.dr.read();
+        let year = bcd2_decode(dr.yt().bits(), dr.yu().bits());
+        Ok(year as u16)
+    }
+
+    fn get_date(&mut self) -> Result<NaiveDate, Self::Error> {
+        let day = self.get_day().unwrap();
+        let month = self.get_month().unwrap();
+        let year = self.get_year().unwrap();
+
+        Ok(NaiveDate::from_ymd(year.into(), month.into(), day.into()))
+    }
+
+    fn get_datetime(&mut self) -> Result<NaiveDateTime, Self::Error> {
+        self.set_24h_fmt();
+
+        let day = self.get_day().unwrap();
+        let month = self.get_month().unwrap();
+        let year = self.get_year().unwrap();
+
+        let seconds = self.get_seconds().unwrap();
+        let minutes = self.get_minutes().unwrap();
+        let hours = hours_to_u8(self.get_hours().unwrap());
+
+        Ok(
+            NaiveDate::from_ymd(year.into(), month.into(), day.into()).and_hms(
+                hours.into(),
+                minutes.into(),
+                seconds.into(),
+            ),
+        )
+    }
+}
+
+// Two 32-bit registers (RTC_TR and RTC_DR) contain the seconds, minutes, hours (12- or 24-hour format), day (day
+// of week), date (day of month), month, and year, expressed in binary coded decimal format
+// (BCD). The sub-seconds value is also available in binary format.
+//
+// The following helper functions encode into BCD format from integer and
+// decode to an integer from a BCD value respectively.
+fn bcd2_encode(word: u32) -> (u8, u8) {
+    (
+        (word / 10).try_into().unwrap(),
+        (word % 10).try_into().unwrap(),
+    )
+}
+
+fn bcd2_decode(fst: u8, snd: u8) -> u32 {
+    (fst * 10 + snd).into()
+}
+
+fn hours_to_register(hours: Hours) -> Result<(u8, u8), Error> {
+    match hours {
+        Hours::H24(h) => Ok(bcd2_encode(h as u32)),
+        Hours::AM(h) => Ok(bcd2_encode((h - 1) as u32)),
+        Hours::PM(h) => Ok(bcd2_encode((h + 11) as u32)),
+    }
+}
+
+fn hours_to_u8(hours: rtcc::Hours) -> u8 {
+    if let rtcc::Hours::H24(h) = hours {
+        h
+    } else {
+        panic!("hours could not be destructured into rtc::Hours::H24(h)");
+    }
+}
+
+/// Enable the low frequency external oscillator. This is the only mode currently
+/// supported, to avoid
+fn enable_lse(bdcr: &mut BDCR, bypass: bool) {
+    bdcr.bdcr()
+        .modify(|_, w| w.lseon().set_bit().lsebyp().bit(bypass));
+    while bdcr.bdcr().read().lserdy().bit_is_clear() {}
+}
+
+fn unlock(apb1: &mut APB1) {
+    let pwr = unsafe { &(*PWR::ptr()) };
```

Comment:
This doesn't look safe to me. There is nothing to ensure we have exclusive access to the PWR register here.
Easiest way to solve this would be to pass in PWR as a `&mut`, same as APB1.